collection/collection.go

```diff
@@ -0,0 +1,247 @@
+	}
+
+	results := make([]dsref.VersionInfo, 0, lp.Limit)
+
+	for _, item := range col {
+		lp.Offset--
+		if lp.Offset > 0 {
+			continue
+		}
+
+		results = append(results, item)
+	}
+
+	return results, nil
+}
+
+func (c *collection) Put(ctx context.Context, pid profile.ID, items ...dsref.VersionInfo) error {
+	c.Lock()
+	defer c.Unlock()
+
+	for _, item := range items {
+		if err := c.putOne(pid, item); err != nil {
+			return err
+		}
+	}
+
+	agg, _ := dsref.NewVersionInfoAggregator([]string{"name"})
+	agg.Sort(c.collections[pid])
+
+	return c.saveProfileCollection(pid)
+}
+
+func (c *collection) putOne(pid profile.ID, item dsref.VersionInfo) error {
+	if item.ProfileID == "" {
+		return fmt.Errorf("profileID is required")
+	}
+	if item.InitID == "" {
+		return fmt.Errorf("initID is required")
+	}
+	if item.Username == "" {
+		return fmt.Errorf("username is required")
+	}
+	if item.Name == "" {
+		return fmt.Errorf("name is required")
+	}
+
+	c.collections[pid] = append(c.collections[pid], item)
+	return nil
+}
+
+func (c *collection) Delete(ctx context.Context, pid profile.ID, initID ...string) error {
+	c.Lock()
+	defer c.Unlock()
+
+	col, ok := c.collections[pid]
+	if !ok {
+		return fmt.Errorf("no collection for profile")
+	}
+
+	for _, removeID := range initID {
+		found := false
+		for i, item := range col {
+			if item.InitID == removeID {
+				found = true
+				copy(col[i:], col[i+1:])              // Shift a[i+1:] left one index.
+				col[len(col)-1] = dsref.VersionInfo{} // Erase last element (write zero value).
+				col = col[:len(col)-1]                // Truncate slice.
+				break
+			}
+		}
+
+		if !found {
+			fmt.Println("can't find id", removeID)
+			return fmt.Errorf("no dataset in collection with initID %q", removeID)
+		}
+	}
+
+	c.collections[pid] = col
+
+	return c.saveProfileCollection(pid)
+}
+
+func (c *collection) loadAll() error {
+	f, err := os.Open(c.basePath)
+	if err != nil {
+		return err
+	}
+
+	names, err := f.Readdirnames(-1)
+	if err != nil {
+		return err
+	}
+
+	c.collections = make(map[profile.ID][]dsref.VersionInfo)
+
+	for _, filename := range names {
+		if isCollectionFilename(filename) {
+			if err := c.loadProfileCollection(filename); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *collection) loadProfileCollection(filename string) error {
+	pid, err := profile.IDB58Decode(strings.TrimSuffix(filename, ".json"))
+	if err != nil {
+		return fmt.Errorf("decoding profile ID: %w", err)
+	}
+
+	f, err := os.Open(filepath.Join(c.basePath, filename))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	items := []dsref.VersionInfo{}
+	if err := json.NewDecoder(f).Decode(&items); err != nil {
+		return err
+	}
+
+	c.collections[pid] = items
+	return nil
+}
+
+func (c *collection) saveProfileCollection(pid profile.ID) error {
+	if c.basePath == "" {
+		return nil
+	}
+
+	items := c.collections[pid]
+	if items == nil {
+		return fmt.Errorf("cannot save empty collection")
```

Comment:
Error is a bit misleading. This isn't an empty collection (which a user may end up at if they Put a ref then Delete it), it's a user whose pid is not in the map, so a totally unknown user.